src/basic/namespace-util.c

```diff
@@ -210,14 +213,40 @@ int userns_acquire(const char *uid_map, const char *gid_map) {
 
         /* Forks off a process in a new userns, configures the specified uidmap/gidmap, acquires an fd to it,
          * and then kills the process again. This way we have a userns fd that is not bound to any
-         * process. We can use that for file system mounts and similar. */
+         * process. We can use that for file system mounts and similar.
+         * If the owner of the new namespace is specified, setuid to it in the child. Use a pipe to sync and
+         * ensure the mapping is done after setuid(). */
 
-        r = safe_fork("(sd-mkuserns)", FORK_CLOSE_ALL_FDS|FORK_DEATHSIG|FORK_NEW_USERNS, &pid);
+        if (pipe2(pipe_fd, 0) < 0)
+                return log_error_errno(errno, "Failed to open bidirectional pipe: %m");
+
+        /* If we have to set a different owner we have to setuid before unsharing the new namespace, otherwise
+         * the kernel will refuse the operation. */
+        if (owner == UID_INVALID)
+                flags |= FORK_NEW_USERNS;
```

Comment:
I'm sorry I understand this less and less. Why do you raise `FORK_NEW_USERNS` when the uid is invalid? The explanation on top doesn't make sense to me.

Reply:
Ok, I see. It's for the case where we don't have to have the userns owned by a specific uid. Sorry, I missed that part

Reply:
Though I'd rephrase that comment. As it stands the code documents the code in the child setuid()+unshare(CLONE_NEWUSER) and not the code that it's on top of.